### PR TITLE
Adding CPU and I/O metrics to process module

### PR DESCRIPTION
**Pull request by juturu**

### CODE REVIEW

File: spec/api-process-spec.js [hunk 1]
```diff
@@ -11,6 +11,9 @@ describe('process module', function () {
 
   describe('process.getIOCounters()', function () {
     it('returns an io counters object', function () {
+      if (process.platform !== 'win32') {
+        return;
```
Comment (zeke) [inline spec/api-process-spec.js]: Heads up: that semicolon is going to break the linter.

File: atom/common/api/atom_bindings.cc [hunk 2]
```diff
@@ -68,6 +69,38 @@ v8::Local<v8::Value> GetSystemMemoryInfo(v8::Isolate* isolate,
...
+}
+
+v8::Local<v8::Value> GetIOCounters(v8::Isolate* isolate) {
+  std::unique_ptr<base::ProcessMetrics> metrics(
+      base::ProcessMetrics::CreateCurrentProcessMetrics());
+  base::IoCounters io_counters;
+  const bool got_counters = metrics->GetIOCounters(&io_counters);
+  mate::Dictionary dict = mate::Dictionary::CreateEmpty(isolate);
+
+  if (got_counters) {
```
Comment (kevinsawicki) [inline atom/common/api/atom_bindings.cc]: I think it is fine to just right this as:

```cpp
if (metrics->GetIOCounters(&io_counters))
```

And then remove the `got_counters` variable.

File: spec/api-process-spec.js (hunk 1 above)
Comment (juturu) [inline spec/api-process-spec.js]: taken care

File: atom/common/api/atom_bindings.cc (hunk 2 above)
Comment (juturu) [inline atom/common/api/atom_bindings.cc]: taken care!

File: spec/api-process-spec.js [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+      assert.equal(typeof cpuUsage.idleWakeupsPerSecond, 'number')
+    })
+  })
+
+  describe('process.getIOCounters()', function () {
+    it('returns an io counters object', function () {
+      if (process.platform !== 'win32') {
+        return
+      }
+      var ioCounters = process.getIOCounters()
```
Comment (kevinsawicki) [inline spec/api-process-spec.js]: `var` -> `const`

File: spec/api-process-spec.js [hunk 4]
```diff
@@ -0,0 +1,26 @@
+const assert = require('assert')
+
+describe('process module', function () {
+  describe('process.getCPUUsage()', function () {
+    it('returns a cpu usage object', function () {
+      var cpuUsage = process.getCPUUsage()
```
Comment (kevinsawicki) [inline spec/api-process-spec.js]: `var` -> `const`

File: docs/api/process.md [hunk 5]
```diff
@@ -116,3 +116,15 @@ Returns `Object`:
...
 Returns an object giving memory usage statistics about the entire system. Note
 that all statistics are reported in Kilobytes.
+
+### `process.getCPUUsage()`
+
+Returns:
+
+* `CPUUsage` [CPUUsage](structures/cpu-usage.md)
+
+### `process.getIOCounters()` _Windows_ _Linux_
```
Comment (kevinsawicki) [inline docs/api/process.md]: Does this just return an empty object on macOS then?

File: docs/api/structures/cpu-usage.md [hunk 6]
```diff
@@ -0,0 +1,6 @@
+# CPUUsage Object
+
+* `percentCPUUsage` number - Percentage of CPU used since the last call to getCPUUsage.
```
Comment (kevinsawicki) [inline docs/api/structures/cpu-usage.md]: `number` -> `Number`

File: docs/api/structures/cpu-usage.md [hunk 7]
```diff
@@ -0,0 +1,6 @@
...
+  First call returns 0.
+* `idleWakeupsPerSecond` number - The number of average idle cpu wakeups per second
```
Comment (kevinsawicki) [inline docs/api/structures/cpu-usage.md]: `number` -> `Number`

File: docs/api/process.md (hunk 5 above)
Comment (juturu) [inline docs/api/process.md]: Yes, returns empty on mac. see https://cs.chromium.org/chromium/src/base/process/process_metrics_mac.cc?q=GetIOCounters+package:%5Echromium$&dr=CSs&l=400

File: spec/api-process-spec.js [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+    it('returns a cpu usage object', function () {
+      const cpuUsage = process.getCPUUsage()
+      assert.equal(typeof cpuUsage.percentCPUUsage, 'number')
+      assert.equal(typeof cpuUsage.idleWakeupsPerSecond, 'number')
+    })
+  })
+
+  describe('process.getIOCounters()', function () {
+    it('returns an io counters object', function () {
+      if (process.platform !== 'win32') {
```
Comment (kevinsawicki) [inline spec/api-process-spec.js]: Should this be `process.platform == 'darwin'` since this spec should work on Linux as well?

**Comment by kevinsawicki** [comment]
Thanks for this @juturu 🔍 👍

**Comment by zeke** [comment]
This is great, @juturu.

Would you be interested in collaborating with me on a short guide to using these new APIs as well as the `app.getAppMemoryInfo()` method you just added in https://github.com/electron/electron/pull/9214?

**Comment by juturu** [comment]
Good idea @zeke. Will try to squeeze this in and report back (may take a week).

**Pull request merged**